### ISSUE TITLE
Script scheduling

### ISSUE THREAD

**Issue by cgb** [Ylianst/MeshCentral#4444]
Hey @Ylianst thank you for the addition of script running, it seems to be working well and I like the new drag and drop feature also :)

One thing I've been using Scripttask for is to schedule commands to run periodically, and collect/analyse the output as it is stored in the database.

Was wondering if that you had thought about integrating that feature into the new script run capabilities?

**Comment by Ylianst** [comment]
Yes! I plan to do that. I first want to support getting back the result of a script and storing the result in the event log and display that back. I will then work on script scheduling.

I am on vacation September 7th to October 17th so, I am not sure what the timing for the releases of these feature will be. But that is certainly the plan.

**Comment by cgb** [comment]
Awesome thank you, completely understand.  I logged the issue as I wasn't sure if you were familiar with the scheduling part of Scripttask and maybe your scope was just a interface for running scripts :)

Enjoy your vacation and looking forward to seeing an update in a few months

**Comment by Ylianst** [comment]
Logging this issue also shows you care about this feature. Nothing worst than coding something no-one will use. So, totally appreciate you taking the time to file it. I am working on testing the new version of the MeshAgent today, if it passes, that will be a big deal. The new agent is much better.

**Comment by warlord1312** [comment]
> Yes! I plan to do that. I first want to support getting back the result of a script and storing the result in the event log and display that back. I will then work on script scheduling.
> 
> I am on vacation September 7th to October 17th so, I am not sure what the timing for the releases of these feature will be. But that is certainly the plan.

Would it be possible to store the result in the details tab in a custom field for each script? In the future it would make it easier to export things into a nice report as well versus having it as an event log.

**Comment by cgb** [comment]
@warlord1312 I analyse the output of the same script running multiple times, so I think each command will need an event timestamp along with other metadata.

For comparison, an example job output from Scripttask:
```json

{
  "_id": {
    "$oid": "XXX",
  },
  "node": "node//XXX",
  "type": "job",
  "runBy": "username",
  "errorVal": null,
  "scriptId": "XXX",
  "queueTime": 1648768204,
  "returnAct": null,
  "returnVal": "C:\\Program Files\\Mesh Agent>ping -n 10 -w 1000 1.1.1.1 \r\n\r\nPinging 1.1.1.1 with 32 bytes of data :\r\nReply from 1.1.1.1: bytes=32 time=550ms TTL=58\r\nRequest timed out.\r\nReply from 1.1.1.1: bytes=32 time=428ms TTL=58\r\nReply from 1.1.1.1: bytes=32 time=511ms TTL=58\r\nReply from 1.  1.1.1: bytes=32 time=325ms TTL=58\r\nRequest timed out.\r\nRequest timed out.\r\nRequest timed out.\r\nRequest timed out.\r\nRequest timed out.\r\n\r\nPing statistics for 1.1.1.1:\r\n    Pac kets: Sent = 10, Received = 4, Lost = 6 (60% loss),\r\nApproximate round trip times in milli-seconds:\r\n    Minimum = 325ms, Maximum = 550ms, Average = 453ms",
  "scriptName": "ping-1111",
  "jobSchedule": {
    "$oid": "XXX",
  },
  "replaceVars": null,
  "completeTime": 1652765314,
  "dispatchTime": 1652765294,
  "dontQueueUntil": 1648768140
}
```

**Comment by warlord1312** [comment]
@cgb Can you expand on your exact use case?

I do not want to go off topic here but it may also be helpful to have an alert being sent based on the result of the script maybe?

**Comment by crettien** [comment]
> Logging this issue also shows you care about this feature. Nothing worst than coding something no-one will use. So, totally appreciate you taking the time to file it. I am working on testing the new version of the MeshAgent today, if it passes, that will be a big deal. The new agent is much better.

Hello Ylian, this feature is also of great interest to me and I'll definitely be using it. Thanks for your excellent work!

**Comment by ne0YT** [comment]
@Ylianst hey there, any update on this ? :)
It's really the last part I miss from this software to make it a perfect alternative for so many tools..

**Comment by zekecato** [comment]
I'm very interested in this feature as well.

**Comment by silversword411** [comment]
Official Development has stopped on MeshCentral, unless you or someone else is going to develop and release it yourself I wouldn't expect it. https://github.com/Ylianst/MeshCentral/discussions/4795

If you want a script scheduling mechanism, Tactical RMM which works with MeshCentral is an option https://github.com/amidaware/tacticalrmm

**Comment by DaanSelen** [comment]
Is there any update @Ylianst ? I am not very used to github work so excuse my (maybe unnecessary) mention.

**Comment by DaanSelen** [comment]
https://github.com/meshcentral-extensions/MeshCentral-ScriptTask

**Comment by ne0YT** [comment]
> https://github.com/meshcentral-extensions/MeshCentral-ScriptTask

This is sadly not looking/working how I would like it.
I know that if Ylianst would have implemented the functions it would be perfect.

**Comment by DaanSelen** [comment]
> > https://github.com/meshcentral-extensions/MeshCentral-ScriptTask
> 
> This is sadly not looking/working how I would like it. I know that if Ylianst would have implemented the functions it would be perfect.

What are you missing? We always welcome volunteers for PR's or/and suggestions.

**Comment by ne0YT** [comment]
> > > https://github.com/meshcentral-extensions/MeshCentral-ScriptTask
> > 
> > 
> > This is sadly not looking/working how I would like it. I know that if Ylianst would have implemented the functions it would be perfect.
> 
> What are you missing? We always welcome volunteers for PR's or/and suggestions.

The plugin just doesn't work too well..
1 - you can't scroll on the left side (scripts-selection) limiting the max amount of scripts.
https://github.com/ryanblenis/MeshCentral-ScriptTask/issues/25
2 - when I select "Advanced run" on the right hand side I cannot select any host, only when I reload the page it works (I need to do this each time)
3 - it would be nice to be able to select device groups (similar to folders while also being able to select all the devices of a group at once) instead of just displaying all devices seperately
4 - Repeating Scripts being saved in the Mesh Agent folder and filling up the client disk:
https://github.com/ryanblenis/MeshCentral-ScriptTask/issues/46
5 - Users can create and run scripts but have no output in History / only allow a specific group to use this
https://github.com/ryanblenis/MeshCentral-ScriptTask/issues/11

**Comment by ne0YT** [comment]
@Ylianst can you re-open?

**Comment by DaanSelen** [comment]
@si458 do we reopen or mark as not planned.

**Comment by ne0YT** [comment]
@si458 let's reopen another time?

**Comment by DaanSelen** [comment]
@si458 we are probably going to have to decide if we are going to mark this out-of-scope or on the roadmap?

**Comment by si458** [comment]
closing in favor of #3613
